docs: refine Application Command docs to be more concise and organized (#199)
sapphiredev/website: docs/Guide/application-commands/application-command-registry/registering-chat-input-commands.mdx

title: Registering Chat Input Commands
---

> :::warning

Reviewer: ```suggestion
:::caution
```

`waning` becomes a red block
![image](https://user-images.githubusercontent.com/4019718/187927333-d2a8732f-85e0-41f9-b261-3f82f2ba9e8b.png)

To register a Chat Input Command (also known as a Slash Command) with Discord, you need to acquire an application
command registry. If you missed how to do it, please see the [`Acquiring an Application Command Registry`][aaacr]
section for more details.

:::

## How to register a Chat Input Command

> Register a Chat Input Command by calling the [`registerChatInputCommand`][rcic] method on the registry, this looks like:

Reviewer: I'd like this to be reverted

> ```typescript ts2esm2cjs|{7,12-16}|{7,12-16}

Reviewer: revert

:::info

We're using `//` next to `builder //` to prevent Prettier from formatting everything on the same line. If you do not use

> Prettier or don't care, you can remove this.

Reviewer: ```suggestion
Prettier, or don't care, you can remove this.
```

We're using `//` next to `builder //` to prevent Prettier from formatting everything on the same line. If you do not use
Prettier or don't care, you can remove this.

:::

In this code snippet we have registered the Chat Input Command in the most simple way referencing the already defined
`name` and `description`. We also use the [`builder`][djs-slash-builder-docs] pattern which means you get verification
of your input before a command gets registered in a faulty way. It will also make sure that you are not providing
unnecessary or incorrect JSON keys because you're not building the JSON yourself. All that said, it is the method we
recommend everyone to use. If you are experienced with Chat Input Commands you can also opt to write your own raw JSON
object, however this guide will not cover that.

In the code snippet above we passed 1 parameter to the [`registerChatInputCommand`][rcic] method, to build the object
that gets send to Discord for registering the command. The [`registerChatInputCommand`][rcic] however also takes a

> second parameter which we will cover next.

Reviewer: revert, being more verbose is better.

}
```

In this example, we will add an option that will ask for another user as input. The option can be configured with
`setRequired(true)`, which means Discord natively prevents the chat input command from ever being run if it isn't
provided. This is really nice compared to using Sapphire's [Arguments][arguments] for input validation in message
commands.

There are many other types of options that you can add to your Chat Input Command, and you can read all about them on
the [`discord.js guide`][djs-slash-command-options].

## Chat Input Command Registry Options

These are options that we can pass into the [`registerChatInputCommand`][rcic] method as a second argument, not to be
confused with a Chat Input Command's options in the previous section. Let's take a look at them below!

### [`idHints`][idhints]

At a high level, an application command registry maps application command IDs and/or names to your command. By providing

> ID(s) to the `idHints` option, we can prevent new commands from accidentally being created.

Reviewer: ```suggestion
ID(s) to the [`idHints`][idhints] option, we can prevent new commands from accidentally being created.
```

[Discord's API directly][discord-deleting-application-commands], but it's better to prevent this in the first place!

The first time an application command is registered, Sapphire will log this ID at the [`info`][loglevel-info] level to
the console. We **highly recommend** you copy this ID and add it to `idHints` option array. If you miss it the first
time, Sapphire will continue to log this ID at the [`debug`][loglevel-debug] level. If you have configured Sapphire's
Logger to only log errors or fatal messages, please check [`Configuring Log Level Guide`][configure-logging] to learn
how to change the log level visibility.

:::

### [`guildIds`][guildids]

Discord allows you to register application commands either globally or for specific guilds. Providing guild ids to
`guildIds` will only register the application command to those guilds.

By default, this is `undefined` which registers the application command globally. However, an empty array acts the same
way which may be useful if you're fetching guild ids from another source (i.e. an external API or database).

### [`behaviorWhenNotIdentical`][behaviorwhennotidentical]

> This options tells the registry that how to behave when the command is not identical to the one which is already

Reviewer: ```suggestion
This option tells the registry how to behave when the command is not identical to the one which is already
```
